### ISSUE TITLE
Usage example notebook (`examples/knowledge_distillation.ipynb`)

### ISSUE THREAD

**Issue by peremartra** [peremartra/optipfair#26]
# Sub-issue 5/5: Usage example notebook (`examples/knowledge_distillation.ipynb`)

**Parent issue:** #12  
**Depends on:** Sub-issues 1, 2, 3 and 4 (full distillation module must be complete and integrated)

---

## Context

This task creates the usage example notebook for the distillation module. It is a direct adaptation of `CH06_NB03_Hands_on_qwen3_5.ipynb`, rewritten to use OptiPFair's public API instead of the raw training functions defined inside the notebook. Benchmark evaluations with `lm_eval` are excluded — the notebook covers only the pruning and knowledge distillation steps.

The notebook demonstrates the full recommended workflow:
1. Load teacher model
2. Prepare dataset
3. Create a depth-pruned student using OptiPFair
4. Recover performance via knowledge distillation using `opf.distill_model()`
5. Visualize training loss curves

---

## Files to Create
```
examples/knowledge_distillation.ipynb
```

---

## Notebook Structure
use a header similar to the one in: data-driven_width_pruning.ipynb with the image and the social links. 
Add also a explanation about the notebook. 
### Section 0 — Environment & Dependencies
```python
# Install
!pip install optipfair transformers datasets tqdm matplotlib

# Constants
RECOVERY_SAMPLES = 2000   # Use a small number for the example
EPOCHS = 3
LEARNING_RATE = 4e-5
BATCH_SIZE = 4
MAX_LENGTH = 512
LAYERS_TO_REMOVE_COUNT = 2
```
```python
# Imports
import torch
from copy import deepcopy
from transformers import AutoModelForCausalLM, AutoTokenizer
from datasets import load_dataset, Dataset
from torch.utils.data import TensorDataset, DataLoader, random_split
from tqdm.auto import tqdm
import matplotlib.pyplot as plt
import optipfair as opf

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Using device: {device}")
```

---

### Section 1 — Load Teacher Model
```python
MODEL_NAME = "Qwen/Qwen3.5-0.8B-Base"   # or "google/gemma-3-270m" for a smaller option

print(f"Loading Teacher model: {MODEL_NAME}")
teacher_model = AutoModelForCausalLM.from_pretrained(
    MODEL_NAME,
    torch_dtype=torch.bfloat16 if torch.cuda.is_available() else torch.float32,
    device_map="auto" if torch.cuda.is_available() else None
)

# Freeze teacher — it is never updated
teacher_model.eval()
for param in teacher_model.parameters():
    param.requires_grad = False

tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
if tokenizer.pad_token is None:
    tokenizer.pad_token = tokenizer.eos_token

n_teacher_layers = teacher_model.config.num_hidden_layers
print(f"✓ Teacher: {n_teacher_layers} layers, {teacher_model.num_parameters():,} params")
```

---

### Section 2 — Prepare Training Dataset

Adapted from NB03. Uses Cosmopedia in streaming mode, tokenizes, and creates an 80/20 train/val split.
```python
# Load Cosmopedia dataset (streaming)
print("Loading Cosmopedia dataset...")
dataset_name = "HuggingFaceTB/cosmopedia"
subsets = ["stories", "wikihow", "openstax", "web_samples_v1"]
samples_per_subset = int(RECOVERY_SAMPLES / len(subsets))

all_samples = []
for subset in subsets:
    print(f"  Loading {subset}...")
    subset_data = load_dataset(dataset_name, subset, split="train", streaming=True)
    subset_samples = list(subset_data.take(samples_per_subset))
    all_samples.extend(subset_samples)
    print(f"    ✓ {len(subset_samples):,} samples")

distillation_dataset = Dataset.from_dict({'text': [s['text'] for s in all_samples]})
print(f"✓ Total samples: {len(distillation_dataset):,}")
```
```python
# Tokenize
print("Tokenizing...")
texts = [item['text'] for item in distillation_dataset]
tokenized_data = []
for i in tqdm(range(0, len(texts), 100), desc="Tokenizing"):
    batch = tokenizer(
        texts[i:i+100],
        truncation=True,
        padding="max_length",
        max_length=MAX_LENGTH,
        return_tensors="pt"
    )
    tokenized_data.append(batch)

input_ids = torch.cat([b['input_ids'] for b in tokenized_data], dim=0)
attention_mask = torch.cat([b['attention_mask'] for b in tokenized_data], dim=0)
full_dataset = TensorDataset(input_ids, attention_mask)

# 80/20 split
generator = torch.Generator().manual_seed(42)
train_size = int(0.8 * len(full_dataset))
val_size = len(full_dataset) - train_size
train_dataset, val_dataset = random_split(
    full_dataset, [train_size, val_size], generator=generator
)

train_dataloader = DataLoader(train_dataset, batch_size=BATCH_SIZE, shuffle=True)

print(f"✓ Train: {len(train_dataset):,} samples ({len(train_dataloader):,} batches)")
print(f"✓ Val:   {len(val_dataset):,} samples")
```

---

### Section 3 — Create Pruned Student Model
```python
# Step 1: Analyze layer importance using calibration data
print("Analyzing layer importance...")
student_model = deepcopy(teacher_model)
importance_scores = opf.analyze_layer_importance(
    student_model,
    train_dataloader,
    show_progress=True
)

print("\nLayer importance scores (lower = less important):")
for layer_idx, score in sorted(importance_scores.items()):
    print(f"  Layer {layer_idx:2d}: {score:.6f}")
```
```python
# Step 2: Select the N least important layers to remove
LAYERS_TO_REMOVE = sorted(
    importance_scores.keys(),
    key=lambda x: importance_scores[x]
)[:LAYERS_TO_REMOVE_COUNT]

print(f"Layers selected for removal: {LAYERS_TO_REMOVE}")
```
```python
# Step 3: Prune the student via depth pruning
student_model = opf.prune_model_depth(
    model=student_model,
    layer_indices=LAYERS_TO_REMOVE,
    show_progress=True
)

# Unfreeze student — inherited requires_grad=False from the frozen teacher deepcopy
for param in student_model.parameters():
    param.requires_grad = True

n_student_layers = student_model.config.num_hidden_layers
print(f"\n✓ Teacher layers: {n_teacher_layers}")
print(f"✓ Student layers: {n_student_layers} (removed {LAYERS_TO_REMOVE})")
print(f"✓ Student params: {student_model.num_parameters():,}")
```

---

### Section 4 — Knowledge Distillation with OptiPFair

Labels-only training: hard labels + Skew KLD. No feature alignment. This replicates the NB03 training configuration using OptiPFair's `distill_model()`.
```python
from copy import deepcopy

student_to_train = deepcopy(student_model)

trained_student, stats = opf.distill_model(
    student_model=student_to_train,
    teacher_model=teacher_model,
    dataloader=train_dataloader,
    # Loss configuration — Labels Only (default)
    alpha=0.6,          # Hard labels weight: learns from ground truth (dataset)
    beta=0.4,           # Soft labels weight: learns from teacher's output distribution
    gamma=0.0,          # Feature alignment: OFF
    delta=0.0,          # Feature dynamics: OFF
    temperature=2.0,
    skew_alpha=0.4,
    # Training configuration
    epochs=EPOCHS,
    learning_rate=LEARNING_RATE,
    accumulation_steps=4,
    show_progress=True,
    return_stats=True,
)

print(f"\n✓ Training complete")
print(f"  Total time:        {stats['total_time_seconds']:.1f}s "
      f"({stats['total_time_seconds']/60:.1f} min)")
print(f"  Avg time/epoch:    {stats['avg_time_per_epoch']:.1f}s")
print(f"  Final total loss:  {stats['loss_history']['total'][-1]:.4f}")
print(f"  Final task loss:   {stats['loss_history']['task'][-1]:.4f}")
print(f"  Final logits loss: {stats['loss_history']['logits'][-1]:.4f}")
```

---

### Section 5 — Visualize Training Loss Curves

Adapted directly from NB03 Section 6. Uses `stats['loss_history']` from `return_stats=True`.
```python
fig, axes = plt.subplots(1, 3, figsize=(15, 4))

loss_keys = [
    ('total',  'Total Loss'),
    ('task',   'Task Loss (Cross-Entropy)'),
    ('logits', 'Logits Loss (Skew KLD)'),
]

for idx, (key, title) in enumerate(loss_keys):
    axes[idx].plot(
        stats['loss_history'][key],
        marker='o', linewidth=2, markersize=8
    )
    axes[idx].set_title(title, fontsize=12)
    axes[idx].set_xlabel('Epoch')
    axes[idx].set_ylabel('Loss')
    axes[idx].grid(True, alpha=0.3)

plt.suptitle(
    'Training Progress: Knowledge Distillation (Labels Only)',
    fontsize=14, fontweight='bold'
)
plt.tight_layout()
plt.savefig('kd_training_curves.png', dpi=150, bbox_inches='tight')
plt.show()

print(f"✓ Plot saved to kd_training_curves.png")
```

---

### Section 6 — (Optional) Save Trained Student
```python
OUTPUT_PATH = "./kd-student"
trained_student.save_pretrained(OUTPUT_PATH)
tokenizer.save_pretrained(OUTPUT_PATH)
print(f"✓ Trained student saved to {OUTPUT_PATH}")
```

---

## Style Guidelines

Follow the conventions of the existing `examples/depth_pruning.ipynb`:
- Each section begins with a Markdown cell explaining what it does and why
- Code cells are short and focused on one task
- All `print()` statements show key results clearly
- Comments in English

---

## Key Differences with NB03

| Aspect | NB03 | This notebook |
|--------|------|---------------|
| Training function | `train_student_advanced()` defined inline | `opf.distill_model()` |
| Loss function | `compute_compound_loss_advanced()` defined inline | Internal to OptiPFair |
| Stats access | `history` dict from return value | `stats['loss_history']` from `return_stats=True` |
| Benchmarks | `lm_eval` on arc_easy, winogrande, etc. | **Not included** |
| Dataset size | 2000 samples (configurable via `RECOVERY_SAMPLES`) | Same |

---

## Acceptance Criteria

- [ ] `examples/knowledge_distillation.ipynb` exists
- [ ] Sections 0–6 are present and documented with Markdown cells
- [ ] Uses `opf.analyze_layer_importance()` → `opf.prune_model_depth()` → `opf.distill_model()` pipeline
- [ ] `return_stats=True` is used and `stats['loss_history']` drives the loss curve plots
- [ ] No inline definitions of training or loss functions — everything goes through OptiPFair's API
- [ ] The notebook follows the style of existing `examples/` notebooks
